File: problems/chapter16/miiitomi/006.cpp
#include <iostream>
#include <vector>
using namespace std;
typedef long long ll;

const ll INF = (1LL << 60);

struct Edge {
    int rev, to;
    ll cap;
    Edge(int r, int t, ll c) : rev(r), to(t), cap(c) {}
};

ll rec(vector<vector<Edge>> &G, vector<bool> &seen, int v, ll f) {
    if (seen[v]) return 0LL;
    seen[v] = true;
    if (v == (int)G.size()-1) return f;

    for (auto &e : G[v]) {
        if (e.cap == 0LL) continue;
        ll flow = rec(G, seen, e.to, min(e.cap, f));
        if (flow > 0LL) {
            e.cap -= flow;
            G[e.to][e.rev].cap += flow;
            return flow;
        }
    }
    return 0LL;
}

ll FordFulkerson(vector<vector<Edge>> &G) {
    ll ans = 0LL;
    while (true) {
        vector<bool> seen((int)G.size(), false);
        ll flow = rec(G, seen, 0, INF);
        if (flow == 0) break;
        else ans += flow;
    }
    return ans;
}

int main() {
    int N;
    cin >> N;
    vector<vector<Edge>> G(N+2);
    ll max_point = 0LL;

    for (int i = 1; i <= N; i++) {
        ll a;
        cin >> a;
        if (a <= 0LL) {
            G[0].push_back(Edge((int)G[i].size(), i, -a));
            G[i].push_back(Edge((int)G[0].size()-1, 0, 0LL));
        } else {
            max_point += a;
            G[i].push_back(Edge((int)G[N+1].size(), N+1, a));
            G[N+1].push_back(Edge((int)G[i].size(), i, 0LL));
        }
        for (int j = 2 * i; j <= N; j += i) {
            G[i].push_back(Edge((int)G[j].size(), j, INF));
            G[j].push_back(Edge((int)G[i].size()-1, i, 0LL));
        }
    }

    cout << max_point - FordFulkerson(G) << endl;
}

File: problems/chapter16/mshibatatt/006.py
# https://atcoder.jp/contests/arc085/tasks/arc085_c

import sys
sys.setrecursionlimit(10000)

class FordFulkerson:
    def __init__(self, N):
        self.N = N
        self.G = [[] for i in range(N)]
 
    def add_edge(self, fr, to, cap):
        forward = [to, cap, None]
        forward[2] = backward = [fr, 0, forward]
        self.G[fr].append(forward)
        self.G[to].append(backward)
 
    def add_multi_edge(self, v1, v2, cap1, cap2):
        edge1 = [v2, cap1, None]
        edge1[2] = edge2 = [v1, cap2, edge1]
        self.G[v1].append(edge1)
        self.G[v2].append(edge2)
 
    def dfs(self, v, t, f):
        if v == t:
            return f
        seen = self.seen
        seen[v] = True
        for e in self.G[v]:
            w, cap, rev = e
            if cap and not seen[w]:
                d = self.dfs(w, t, min(f, cap))
                if d:
                    e[1] -= d
                    rev[1] += d
                    return d
        return 0
 
    def flow(self, s, t):
        flow = 0
        f = INF = float('inf')
        N = self.N
        while f:
            self.seen = [False]*N
            f = self.dfs(s, t, INF)
            flow += f
        return flow

def main():
    N = int(input())
    a = list(map(int, input().split()))
    ff = FordFulkerson(N+2)
    
    sum_a = 0
    for i, x in enumerate(a):
        if x < 0:
            ff.add_edge(N, i, abs(x))
            ff.add_edge(i, N+1, 0)
        else:
            ff.add_edge(N, i, 0)
            ff.add_edge(i, N+1, abs(x))
            sum_a += x
   
    for i in range(N):
        for j in range(i+1, N):
            if (j+1) % (i+1) == 0:
                ff.add_edge(i, j, float('inf'))
    
    output = sum_a - ff.flow(N, N+1)
    print(output)

if __name__=='__main__':
    main()